examples/Handler Methods.ipynb

```diff
@@ -0,0 +1,717 @@
+    "def _will_process(self, data_container, context):\n",
+    "    data_container, context = BaseStep._will_process(self, data_container, context)\n",
+    "    return ExpandedDataContainer.create_from(data_container), context"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "3. Implement the _did_process lifecycle method to reduce the dimension of the data inside the data container."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def _did_process(self, data_container, context):\n",
+    "    data_container = BaseStep._did_process(self, data_container, context)\n",
+    "    return data_container.reduce_dim()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The result looks like this: "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class ExpandDim(MetaStepMixin, BaseStep):\n",
+    "    def __init__(self, wrapped: BaseStep):\n",
+    "        BaseStep.__init__(self)\n",
+    "        MetaStepMixin.__init__(self, wrapped)\n",
+    "        \n",
+    "    def _will_process(self, data_container, context):\n",
+    "        data_container, context = BaseStep._will_process(self, data_container, context)\n",
+    "        return ExpandedDataContainer.create_from(data_container), context\n",
+    "\n",
+    "    def _did_process(self, data_container, context):\n",
+    "        data_container = BaseStep._did_process(self, data_container, context)\n",
+    "        return data_container.reduce_dim()\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before testing the step, let's define a quick utility class to print data for this example, and the ones after :"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class PrintData(Identity):   \n",
+    "    def __init__(self, message):\n",
+    "        super().__init__()\n",
+    "        self.message = message\n",
+    "        \n",
+    "    def _did_process(self, data_container, context):\n",
+    "        print('data_inputs {}: {}'.format(self.message, data_container.data_inputs))\n",
+    "        print('expected_outputs {}: {}'.format(self.message, data_container.expected_outputs))\n",
+    "        print('\\n')\n",
+    "        return data_container"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's transform our step :"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "data_inputs before: [0 1]\n",
+      "expected_outputs before: [1 2]\n",
+      "data_inputs inside: [array([0, 1])]\n",
+      "expected_outputs inside: [array([1, 2])]\n",
```

Comment:
Please avoid numpy usage whenever it could be avoided. Here it seems like ExpandDim needn't to be a np step, as it's just a wrapper that adds a list `[ x ]` around the data. I have repeated this to you several times this year, I want to make it clear that whenever possible we should avoid numpy to optimize performances and avoid incompatibilities. Use iterables such as lists instead.